python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java

```diff
@@ -140,16 +195,63 @@ private PythonType resolveReturnType(Map<Symbol, PythonType> createdTypesBySymbo
     if (classSymbol == null) {
       return PythonType.UNKNOWN;
     }
-    if (rootModule == null) {
+    if (createdTypesBySymbol.containsKey(classSymbol)) {
+      return createdTypesBySymbol.get(classSymbol);
+    }
+    String fullyQualifiedName = classSymbol.fullyQualifiedName();
+    if (fullyQualifiedName == null) {
       return convertToType(classSymbol, createdTypesBySymbol);
     }
-    // If the symbol is a built-in symbol and the root module already exists, we search for it instead of recreating the symbol
-    // This is necessary to avoid duplicated symbols in the original ProjectSymbolTable to translate into duplicated PythonType
-    // TODO: SONARPY-2010 to fix this
-    return Optional.ofNullable(classSymbol.fullyQualifiedName())
-      .filter(fqn -> !fqn.contains("."))
-      .flatMap(f -> rootModule.resolveMember(f))
-      .orElseGet(() -> convertToType(classSymbol, createdTypesBySymbol));
+    return resolvePossibleLazyType(createdTypesBySymbol, classSymbol, fullyQualifiedName);
+  }
+
+  PythonType resolvePossibleLazyType(Map<Symbol, PythonType> createdTypesBySymbol, ClassSymbol classSymbol, String fullyQualifiedName) {
+    PythonType currentModule = rootModule;
+    String[] split = fullyQualifiedName.split("\\.");
+    StringBuilder sb = new StringBuilder();
+    for (int i = 0; i < split.length; i++) {
+      String currentElement = split[i];
```

Comment:
I think this logic is somehow implemented in `ProjectLevelTypeTable`, so maybe what we just need is to have a link to the types table in lazy type and call `getType` or `getModule` from for unresolved module and then call resolution consumers. WDYT?

Reply:
I moved the `getType` implementation in `SymbolsModuleTypeProvider` (for the reason highlighted in my other comment: that I'm not sure it's a good idea to pass `ProjectLevelTypeTable` to `SymbolsModuleTypeProvider` then to `LazyType`). We can discuss this further though.